decoder/context.go

```diff
@@ -16,8 +21,79 @@ type DecoderContext struct {
 	// CodeLenses represents a slice of executable lenses
 	// which will be executed in the exact order they're declared
 	CodeLenses []lang.CodeLensFunc
+
+	CompletionHooks        CompletionFuncMap
+	CompletionResolveHooks CompletionResolveFuncMap
 }
 
 func (d *Decoder) SetContext(ctx DecoderContext) {
 	d.ctx = ctx
 }
+
+type CompletionFunc func(ctx context.Context, value cty.Value) ([]Candidate, error)
+type CompletionFuncMap map[string]CompletionFunc
+
+type CompletionResolveFuncMap map[string]CompletionResolveFunc
+type CompletionResolveFunc func(ctx context.Context, unresolvedCandidate UnresolvedCandidate) (*ResolvedCandidate, error)
+
+type Candidate struct {
```

Comment:
I'm half wondering if we should call this `HookCandidate`, but given that we don't have a conflict with another "candidate" name within the package I'm fine with this.

Either way though we should document this via comment to clarify that this represents completion candidates returned from hooks.